### PR TITLE
Fix E2E tests - Track gtag event on specific page

### PR DESCRIPTION
**Pull request by jorgemd24**
### Changes proposed in this Pull Request:

As of now, the `trackGtagEvent` function only checks for the event itself. Yet, there are scenarios where the same event may carry different properties, and within a brief timeframe, the same event can be triggered with varying properties. Since we don't perform an assertion to wait for a request with specific data, these tests could potentially fail.

https://github.com/woocommerce/google-listings-and-ads/blob/78582d6ffab31b5bf3dfcf97e69fcbf623f0b932/tests/e2e/utils/track-event.js#L15-L21

This PR tweaks the `trackGtagEvent` function. It will now wait for a specific event and also provides the capability to verify if the "ecomm_pagetype" data is sent.

### Detailed test instructions:

1. Follow these steps to set up the e2e env: https://github.com/woocommerce/google-listings-and-ads#e2e-testing
2. Run `npm run test:e2e -- ./tests/e2e/specs/gtag-events/gtag-events.test.js` and see that all tests passes.

### Changelog entry

> Dev - E2E tests - Track gtag event on specific page

### CODE REVIEW

**Comment by mikkamp** [comment]
> Yet, there are scenarios where the same event may carry different properties, and within a brief timeframe, the same event can be triggered with varying properties.

Can you clarify which scenario this was observed? Is this where we listen for events across various page views or is it for a single page view?
Sending the same event with different parameters on a single page sounds like something isn't being tracked correctly.

**Comment by jorgemd24** [comment]
Hi @mikkamp, it is because of a timing issue in the E2E tests:

https://github.com/woocommerce/google-listings-and-ads/blob/78582d6ffab31b5bf3dfcf97e69fcbf623f0b932/tests/e2e/specs/gtag-events/gtag-events.test.js#L152-L155

When the `singleProductAddToCart` function is called, it calls the `page_view` event without including `ecomm_pagetype`, now I realized that we changed from `networkidle` to `domcontentloaded`, which means that the waitForRequest function could monitor the incorrect request. See here: https://github.com/woocommerce/google-listings-and-ads/pull/2088/files#diff-ad23d05dc6460c47ba77086291640c8373da27bae0b441f8cf9465032e6079e2L34. 

So basically the issue is that `waitForRequest ` was expecting to catch the request in the cart page instead of product page. 


We opted against using `networkidle` because it's discouraged in the documentation and in general it increases the execution time:

`'networkidle' - DISCOURAGED wait until there are no network connections for at least 500 ms. Don't use this method for testing, rely on web assertions to assess readiness instead.
`

However, I'm not sure in this particular scenario, considering what you mentioned earlier about having only one type of event request per page. In this case, it might be worth considering using 'networkidle.'

**Comment by mikkamp** [comment]
Thanks for the explanation, that's helpful to understand the scenario.

I think I'd agree though that in this case it's almost good to have waitForRequest fail as it highlighted the issue that it's not receiving the right event.

I can understand not wanting to use `networkidle`, is it maybe better to switch to waiting till something unique on the cart / product page loads? So we can be certain we are hooking into the event tracking at the right time?

Just to clarify I think the solution in this PR just makes it slightly less transparent what we are actually testing, because of the change it's not apparent that we want to catch the tracking event from the second page load and avoid the one from the first page. In this case they happen to have different parameters, but in theory there might be scenarios where the event is exactly identical.

**Comment by jorgemd24** [comment]
> I can understand not wanting to use networkidle, is it maybe better to switch to waiting till something unique on the cart / product page loads? So we can be certain we are hooking into the event tracking at the right time?

Yes, I will try to look for something "unique" to see if we can avoid the `networkidle`.

> Just to clarify I think the solution in this PR just makes it slightly less transparent what we are actually testing, because of the change it's not apparent that we want to catch the tracking event from the second page load and avoid the one from the first page. In this case they happen to have different parameters, but in theory there might be scenarios where the event is exactly identical.

Yep, when I initially created the PR, I thought it was more of a race condition problem on the same page. However, later on, I realized it was due to the switch from 'networkidle' to 'domcontentloaded.'

**Comment by jorgemd24** [comment]
@mikkamp, I've made some adjustments to the PR to make it clearer on which page the event should be triggered. Let me know what you think. Thanks!

File: tests/e2e/utils/track-event.js [hunk 1]
```diff
@@ -10,14 +10,27 @@
  *
  * @param {Page} page
  * @param {string} eventName Event name to match.
+ * @param {string|null} URLpath The URL path where the event should be triggered.
```
Comment (mikkamp) [inline tests/e2e/utils/track-event.js]: I don't think we need to over engineer this function but maybe add a better description that it matches the starting part of the path.

For example if we go to checkout: `https://domain.test/checkout/`
When we complete checkout we are redirected to: `https://domain.test/checkout/order-received/<orderId>/`

Both of those would match if URLpath is set to `checkout`

File: tests/e2e/utils/track-event.js [hunk 2]
```diff
@@ -10,14 +10,27 @@
...
  * @return {Promise<Request>} Matching request.
  */
-export function trackGtagEvent( page, eventName ) {
+export function trackGtagEvent( page, eventName, URLpath = null ) {
```
Comment (mikkamp) [inline tests/e2e/utils/track-event.js]: Not sure what the recommended format is for naming, but wouldn't `urlPath` or `URLPath` be the better choice?

File: tests/e2e/utils/track-event.js [hunk 3]
```diff
@@ -10,14 +10,27 @@
...
+			url.includes( eventPath ) &&
+			url.includes( match ) &&
+			( URLpath
+				? url.includes(
+						`url=${ encodeURIComponent(
+							`${ origin }/${ URLpath }`
+						) }`
+				  )
+				: true )
+		);
```
Comment (mikkamp) [inline tests/e2e/utils/track-event.js]: I'm not an expert in JS code but it takes me a while to grok what's being compared here.
I wonder if it would be simpler to convert the request URL to a [URL object](https://developer.mozilla.org/en-US/docs/Web/API/URL) so we can check if the event `searchParams` is set and matches the eventName.

That would also allow for more accurate matching with the exact URL. Looking at the payload of a request I see:
image

So using the `url.includes` might not be very foolproof as it could easily match the referrer URL which would be set to the previous page visited.

**Review by mikkamp — COMMENTED** [review]
When switching to this branch I do get the test to work correctly now:
> GTag events › Cart page view event is sent from the cart page

Although for some reason the following two tests still fail locally (although that seems related to something outside of this PR):
> GTag events › Add to cart event is sent from related product on single product page
> GTag events › Add to cart event is sent from the shop page

As outlined in the comments I think we might want to improve the matching a bit more so we don't run into problems later if we use this function to track other scenarios.

File: tests/e2e/utils/track-event.js (hunk 3 above)
Comment (jorgemd24) [reply tests/e2e/utils/track-event.js]: Thanks, @mikkamp, for the suggestion. I don't believe that scenario would occur because I added the prefix `url=` within `url.includes`. Nonetheless, I've made the switch to using URL parameters for the sake of clarity.

**Comment by jorgemd24** [comment]
@mikkamp Thanks for your suggestions, I addressed all of them.

> Although for some reason the following two tests still fail locally (although that seems related to something outside of this PR):

mmmm, that's weird for me all tests are passing:

image

Could you describe the errors that you are finding?

File: tests/e2e/utils/track-event.js (hunk 3 above)
Comment (mikkamp) [reply tests/e2e/utils/track-event.js]: Ah, thanks for clarifying. I seem to have missed the `url=` prefix you added.

Since we now extract the params does it also mean we can change `eventName === params.get( 'event' )`

**Comment by mikkamp** [comment]
image

Seems the event data isn't submitting the correct product ID.

File: tests/e2e/utils/track-event.js (hunk 3 above)
Comment (jorgemd24) [reply tests/e2e/utils/track-event.js]: I don't think so because the `event` is not a parameter, it is included in the `data` parameter.

image

We could do something like: `params.get( 'data' ).includes(match)` if you believe it's a slightly more clear compared to the current implementation.

**Comment by jorgemd24** [comment]
It seems related to this: https://github.com/woocommerce/google-listings-and-ads/issues/2095 but it was already fixed here: https://github.com/woocommerce/google-listings-and-ads/pull/2096

File: tests/e2e/utils/track-event.js (hunk 3 above)
Comment (mikkamp) [reply tests/e2e/utils/track-event.js]: Ok, that makes sense. I do think the last suggestion you made does make it quite explicit what we are expecting to match (which might help with troubleshooting).

Comment (jorgemd24) [reply tests/e2e/utils/track-event.js]: Yep that sounds right. I made the change here: https://github.com/woocommerce/google-listings-and-ads/pull/2122/commits/6557bd19f84999acae74ccc7c4074bf2b51f2a1b

**Review by mikkamp — APPROVED** [review]
Thanks for the final changes and pointing to the related PR.
It helps if I rebuild the `gtag-events` script and don't load an older copy. With a fresh copy everything passes.
